middle_end/flambda/simplify/simplify_set_of_closures.ml

```diff
@@ -472,13 +472,14 @@ let simplify_function context ~used_closure_vars ~shareable_constants
     in
     let function_decl = FD.update_code_id function_decl new_code_id in
     let function_type =
-      (* When not rebuilding terms we cannot infer function types.  However
-         this shouldn't matter too much since inlining will typically be
-         disabled during such traversals. *)
+      (* When not rebuilding terms we only give a skeleton function type so
+         that we can distinguish between direct and indirect calls. *)
```

Comment:
This isn't quite right.  I would say: "When not rebuilding terms we always give a non-inlinable function type, since the body is not available for inlining, but we would still like to generate direct calls to the function".